### PR TITLE
Exists expression

### PR DESCRIPTION
**Pull request by davidmartos96**
Now that we have subqueries modeled, I think the EXISTS expression would be a good addition.
I'm marking it as a Draft because I'm not sure if you want the function `existsQuery` as top level or if it should be an extension method. `(subquery).exists`.
The docstring is missing as well.
Also, I'm not sure about the `boundVariables` check in the test. I messed around with a very adhoc implementation of subqueries yesterday and I was able to expect it bound to `true` and not 1. Not with your implementation though.

### CODE REVIEW

File: moor/test/expressions/subquery_test.dart [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+    final db = TodoDb();
+    final subquery = db.select(db.users)
+      ..where((tbl) => tbl.isAwesome.equals(true));
+    final existsExpression = existsQuery(subquery);
+
+    final context = GenerationContext.fromDb(db);
+    existsExpression.writeInto(context);
+
+    expect(context.sql, 'EXISTS (SELECT * FROM users WHERE is_awesome = ?)');
+    expect(context.boundVariables, [1]);
```
Comment (simolus3) [inline moor/test/expressions/subquery_test.dart:17]: I have a neat test utility that checks all this at once:

```dart
  expect(existsExpression, generates('EXISTS (SELECT * FROM users WHERE is_awesome = ?)', [1]));
```

You'd have to import `../data/expect_generated.dart`.

File: moor/lib/src/runtime/query_builder/expressions/exists.dart [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    if (_not) {
+      context.buffer.write('NOT ');
+    }
+    context.buffer.write('EXISTS ');
+
+    context.buffer.write('(');
+
+    _select.writeInto(context);
+
+    context.buffer.write(')');
```
Comment (simolus3) [inline moor/lib/src/runtime/query_builder/expressions/exists.dart:23]: I don't think we need empty lines between every line here, one after the `EXISTS` should be good enough


```suggestion
    context.buffer.write('(');
    _select.writeInto(context);
    context.buffer.write(')');
```

File: moor/test/expressions/subquery_test.dart [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+  });
+
+  test('not exists expressions are generated', () {
+    final db = TodoDb();
+    final isInExpression = notExistsQuery(db.select(db.users));
+
+    final context = GenerationContext.fromDb(db);
+    isInExpression.writeInto(context);
+
+    expect(context.sql, 'NOT EXISTS (SELECT * FROM users)');
```
Comment (simolus3) [inline moor/test/expressions/subquery_test.dart:27]: After you added the import, we can also apply it here

```suggestion
    expect(isInExpression, generates('NOT EXISTS (SELECT * FROM users)'));
```

File: moor/lib/src/runtime/query_builder/expressions/expression.dart [hunk 4]
```diff
@@ -119,6 +119,14 @@ abstract class Expression<D> implements FunctionParameter {
   }
 }
 
+Expression<bool> existsQuery(BaseSelectStatement select) {
```
Comment (simolus3) [inline moor/lib/src/runtime/query_builder/expressions/expression.dart:122]: IMO these top-level functions should be declared in `exists.dart`.

In terms of documentation, we might have something like:

```suggestion
/// The `EXISTS` operator checks whether the [select] subquery returns any rows.
Expression<bool> existsQuery(BaseSelectStatement select) {
```

File: moor/lib/src/runtime/query_builder/expressions/expression.dart [hunk 5]
```diff
@@ -119,6 +119,14 @@ abstract class Expression<D> implements FunctionParameter {
...
+  return _ExistsExpression(select, false);
+}
+
+Expression<bool> notExistsQuery(BaseSelectStatement select) {
```
Comment (simolus3) [inline moor/lib/src/runtime/query_builder/expressions/expression.dart:126]:
```suggestion
/// The `NOT EXISTS` operator evaluates to `true` if the [select] subquery does
/// not return any rows.
Expression<bool> notExistsQuery(BaseSelectStatement select) {
```

**Review by simolus3 — COMMENTED** [review]
Thank you for this! I agree that it's a useful application of subqueries in the Dart api.

>  I'm not sure if you want the function `existsQuery` as top level or if it should be an extension method. `(subquery).exists.`

I don't have strong opinions here either. Extension methods can have better discoverability, but exists kind of feels like a top-level predicate, not a method on queries. You can make them an extension if you want, but I'm fine with this approach too.

> I messed around with a very adhoc implementation of subqueries yesterday and I was able to expect it bound to true and not 1

Variables always map to simple values, so expecting it to be `1` seems correct to me.

**Comment by davidmartos96** [comment]
@simolus3 Thanks for the recommendations! I applied the corrections.

**Comment by simolus3** [comment]
Thanks! :+1:

**Pull request merged**